fish: native environment

kevingriffin opened the pull request:
###### Motivation for this change
fish shell startup with `programs.fish.enable` set is about 2x slower than without based on my tests
and reports from others. Tracing reveals calls to `fenv` being the vast majority of the time.

A plugin called babelfish, which I've added as a package in https://github.com/NixOS/nixpkgs/pull/108946, can be used to pre-compute native fish code from bash at system build time, eliminating the need to call `fenv` during each shell launch.

This pull request is based off of #108946, but can be rebased onto master if #108946 is accepted.

I've removed calls to fenv for the init configuration specified on the fish module's configuration, as I couldn't find a reason for these to need translation, since the configuration is local to the fish module itself.

A similar pull request exists for nix-darwin, https://github.com/LnL7/nix-darwin/pull/270, where I've vendored the package until it lands in nixpkgs.


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

toonn commented:
Since you're probably most familiar with the tools rn. Is there a reason `fenv` can't be used to precompute the environment?

pacien commented:
CC @bouk (https://github.com/NixOS/nixpkgs/pull/107834#discussion_r549615685)

bouk commented on nixos/modules/programs/fish.nix:
```diff
@@ -157,11 +138,6 @@ in
-
-        set --prepend fish_function_path ${pkgs.fishPlugins.foreign-env}/share/fish/vendor_functions.d
-        fenv source /etc/fish/foreign-env/interactiveShellInit > /dev/null
-        set -e fish_function_path[1]
```
Same as in [nix-darwin](https://github.com/LnL7/nix-darwin/pull/270/files#r554594991), we need to keep these sourcings but should translate all those init files with babelfish

kevingriffin replied:
Ah, I see, I was thinking `cfge` was the configuration for the module, but it's the environment configuration. I'll make the change to translate these with babelfish here and in nix-darwin.

kevingriffin commented:
@toonn I actually started with that approach [here](https://github.com/kevingriffin/plugin-foreign-env/commit/9e92894586ef3e49d03fbcedfcae311e0e75fada), but decided against it because interpolation was a problem: it's done eagerly, so if your config has `environment.variables.FOO = (tty)`, you'll get something like `(no tty)`as the value, because the subshell was evaluated at precompute time, not runtime as was intended.

cole-h reviewed (changes requested):
While this is intriguing, I'd argue against switching to babelfish as the default. According to the project's own README, there is still [some amount of work left to do](https://github.com/bouk/babelfish#to-do). If anybody depends on any of those not-yet-implemented features (e.g. one of the "bunch of arithmetic expressions"), their system might no longer work as expected.

kevingriffin commented:
@cole-h Thank you for the comment. I'd like to hear from @bouk about what those are specifically and generally how suitable this is. 

Should it not be a suitable default, is there any path forward that allows those for whom this would work to use it? Perhaps some kind of option on the fish module? The speedup in my shell is great, and I'd love to find some way to get it to others.

kevingriffin committed:
fish: add babelfish translation option

Configuration may be ran through fenv at shell start time
(as previously) or translated to fish at build time with
the babelfish package.

kevingriffin commented:
@cole-h I've committed a version with new configuration, defaulted to false, on the fish module that enables this behavior. I'd be grateful if you could take a look.

bouk commented:
I'd say babelfish will work fine for 100% of the use case we have in Nixpkgs. At some point I manually went though everything to make sure it does.

What it doesn't support yet are quite niche and definitely not used in nixpkgs shell inits (things like arithmetic expressions)

cole-h commented on nixos/modules/programs/fish.nix:
```diff
@@ -29,6 +45,14 @@ in
+        type = types.bool;
+        default = false;
+        description = ''
+          Whether to translate the configured environment to native fish.
```
```suggestion
          If enabled, the configured environment will be translated to native fish using <link xlink:href="https://github.com/bouk/babelfish">babelfish</link>. Otherwise, <link xlink:href="https://github.com/oh-my-fish/plugin-foreign-env">foreign-env</link> will be used.
```

kevingriffin replied:
This is a lot clearer. Thank you!

cole-h commented on nixos/modules/programs/fish.nix:
```diff
@@ -29,6 +45,14 @@ in
         type = types.bool;
       };
 
+      translateEnvironment = mkOption {
```
The environment is "translated" either way, it's just what is used to translate it changes. Alternatively, change the type to `types.enum [ "babelfish" "foreign-env" ]`.

```suggestion
      useBabelfish = mkOption {
```

kevingriffin replied:
I can see that. I'll use the enum approach and default to "foreign-env".

kevingriffin replied:
Actually, an enum with two values is pretty weird to switch on (if-else if-else? but the type should make the else useless). I think I'll go with `useBabelfish` unless a third option is added to the module.

toonn replied:
Maybe an attrset is a nice solution?
```
{ "foreign-env" = fenv-case; "babelfish" = babelfish-case; }.${cfg.translator}
```

cole-h commented on nixos/modules/programs/fish.nix:
```diff
@@ -105,72 +129,165 @@ in
+
+        # source the NixOS environment config
+        if [ -z "$__NIXOS_SET_ENVIRONMENT_DONE" ]
+          ${(if cfg.translateEnvironment
+          then "source /etc/fish/set-environment.fish"
+          else "fenv source ${config.system.build.setEnvironment}")}
```
Useless parens.

```suggestion
          ${if cfg.translateEnvironment
          then "source /etc/fish/set-environment.fish"
          else "fenv source ${config.system.build.setEnvironment}"}
```

kevingriffin replied:
I'll incorporate this change.

cole-h commented on nixos/modules/programs/fish.nix:
```diff
@@ -105,72 +129,165 @@ in
+        etc."fish/nixos-env-preinit.fish".text = ''
+        # This happens before $__fish_datadir/config.fish sets fish_function_path, so it is currently
+        # unset. We set it and then completely erase it, leaving its configuration to $__fish_datadir/config.fish
+        set fish_function_path ${pkgs.fishPlugins.foreign-env}/share/fish/vendor_functions.d $__fish_datadir/functions
```
Seems weird to rely on both foreign-env and babelfish if babelfish is enabled. This should probably be gated on babelfish not being used.

kevingriffin replied:
I agree. It's a little tricky because the guard below splicing the setup and teardown makes the function above not applicable, but I'll try to do it as cleanly as possible to avoid the dependency.

cole-h commented on nixos/modules/programs/fish.nix:
```diff
@@ -13,6 +13,22 @@ let
+
+  envInteractiveShellInit = pkgs.writeText "interactiveShellInit" cfge.interactiveShellInit;
+
+  sourceEnv = useTranslated: path:
+  if useTranslated then
```
```suggestion
  sourceEnv = path:
  if cfg.useBabelfish then
```

kevingriffin replied:
Do you prefer internalizing the configuration check? It's true the parameter never takes on another value. I'll make this change.

cole-h commented:
@bouk I'm not only talking about inside Nixpkgs, but also in users' configs. No matter how niche, if a user relies on some unimplemented arithmetic expression in the shell inits in their local configs, using babelfish will break their setup. I'm more in favor of an option where users opt into potential breakage (which, thankfully, @kevingriffin has done here). I'm glad you're interested in maintaining this, however, and look forward to it being feature-complete!

cole-h commented on maintainers/maintainer-list.nix:
```diff
@@ -4666,6 +4666,12 @@
     githubId = 494012;
     name = "Kevin Cox";
   };
+  kevingriffin = {
```
Please split this into a separate commit: `maintainers: add kevingriffin`.

kevingriffin commented:
@cole-h I've added a commit which I believe addresses all of your feedback. If it looks good, I can squash it and port the changes back to nix-darwin.

cole-h commented on nixos/modules/programs/fish.nix:
```diff
@@ -105,72 +130,173 @@ in
+      })
+
+      {
+        etc."fish/nixos-env-preinit.fish".text = ''
+        ${if cfg.useBabelfish
```
Nit: the `''${` is unnecessary -- you can just if-else immediately.

e.g.:

```nix
etc."fish/nixos-env-preinit.fish".text =
  if cfg.useBabelfish then
    ''...''
# ........
```

kevingriffin commented:
I've removed the babelfish package commits and rebased on master to get the package. I also addressed the feedback from @cole-h added afterwards.

pacien commented on nixos/modules/programs/fish.nix:
```diff
@@ -105,72 +130,171 @@ in
+        ${pkgs.babelfish}/bin/babelfish < ${config.system.build.setEnvironment} > $out;
+      '';
+
+       etc."fish/shellInit".source = pkgs.runCommand "shellInit" {
```
Weird indentation here and below. You could also factorise those blocs into a function.

kevingriffin replied:
That's a good observation. I've done that, and in the process, hopefully cleaned up the indentation issue you mentioned. I also took @bouk's [advice](https://github.com/LnL7/nix-darwin/pull/270#discussion_r554938896) and made the files all have a .fish extension.

cole-h reviewed (changes requested):
Sorry for always noticing new things, but I think we're getting there. Thanks for your work on this.

After these suggestions are addressed (either accepted or refuted), I'll build my system against this PR and test functionality to the best of my ability / setup.

bouk commented on nixos/modules/programs/fish.nix:
```diff
@@ -105,72 +135,152 @@ in
+
+      (mkIf (!cfg.useBabelfish)
+      {
+        etc."fish/shellInit".text = cfge.shellInit;
+        etc."fish/loginShellInit".text = cfge.loginShellInit;
+        etc."fish/interactiveShellInit".text = cfge.interactiveShellInit;
```
I don't think we need to keep these under foreign-env, especially since we have the flag now.

kevingriffin replied:
What's the general philosophy with allowing nix to set up `/etc`? Is it considered to be part of the activated system and just linked into `/etc` for compatibility? If so, I sort of want to say that digging into the paths in there is sort of like depending on the internal structure of, say `/run/current-system`. Is that something users expect to be able to do?

I do see the appeal of putting the non-fish files in a directory if the parent is going to be `/etc/fish`. Of course the babelfish-translated files won't go in there, as you suggested.

I don't see why using `.source` with the derivations created up top wouldn't work. That is a bit more consistent, as the configuration in then only consumed once. I'll make that change.

cole-h replied:
@bouk Sorry if I was unclear, but this was just for the bash files that will be translated at runtime to fish via fish-foreign-env. This way, the only thing that changes is if people decide to use babelfish, which will then (of course) change the topography of that directory.

Basically, just ensuring backwards compatibility.

kevingriffin commented on nixos/modules/programs/fish.nix:
```diff
@@ -13,6 +13,27 @@ let
+    '';
+
+  babelfishTranslate = path:
+    pkgs.runCommand "set-environment.fish" {
```
I suppose I could get fancy and try to `basename` the provided path, but that might make the code needlessly hard to read. I'll add an argument.

kevingriffin commented:
@cole-h I've pushed a new commit that addresses your latest feedback. If you find anything in your testing, please let me know. If not, I'll squash it all together. Thank you for taking such a close look!

cole-h commented on nixos/modules/programs/fish.nix:
```diff
@@ -105,72 +135,152 @@ in
+    environment = mkMerge [
+      (mkIf cfg.useBabelfish
+      {
+        etc."fish/set-environment.fish".source = babelfishTranslate config.system.build.setEnvironment "setEnvironment";
```
Nit: either the file should be `setEnvironment.fish`, or the arg to `babelfishTranslate` should be `set-environment`. (This will also need to be changed down below in the `nixos-env-preinit.fish` if you decide to rename the file and not the arg.)

Aside from that, next step is testing all this! If I don't get to it today (I'll try my hardest, I promise), it'll probably be Saturday or Sunday.

kevingriffin replied:
Updated to `setEnvironment`. Thanks for the testing!

cole-h commented:
Works mostly fine from my brief testing (both with and without babelfish), except I'm marking this as blocked by https://github.com/bouk/babelfish/issues/6 and https://github.com/bouk/babelfish/issues/8. Once those are resolved and the next version is packaged and merged, this should be good to go.

bouk commented:
I've released v1.0.0 of babelfish with the fixes :+1:

cole-h commented:
Great, thanks! Now we're just waiting on https://github.com/NixOS/nixpkgs/pull/109541 to be merged. 

@kevingriffin Once that's done, please rebase this PR (so testing is easier e.g. I don't have to cherry-pick that PR locally) and I'll test again.

kevingriffin commented:
@cole-h I just rebased onto master to get the new babelfish package. I also squashed all of my commits into a single commit, just in case this ends up being mergeable once you've done some more testing. Thank you for the help!

kevingriffin commented:
@bouk I'm seeing this in my testing:

```~/src/nixpkgs on fish-native-environment
❯ nix-shell -p hello
warning: Nix search path entry '/nix/var/nix/profiles/per-user/root/channels/nixos ' does not exist, ignoring
error: file 'nixpkgs' was not found in the Nix search path (add it using $NIX_PATH or -I), at (string):1:13

~/src/nixpkgs on fish-native-environment
❯ echo $NIX_PATH
/home/kevin/.nix-defexpr/channels nixpkgs=/nix/var/nix/profiles/per-user/root/channels/nixos  nixos-config=/etc/nixos/configuration.nix  /nix/var/nix/profiles/per-user/root/channels
```

Somewhere an extra space is entering into the equation, but I can't work out where it's coming in. The path in `setEnvironment.fish` seems correct to me. Any ideas? I'll keep testing, but I wanted to put this out there so that it's not merged before this is fixed.

bouk commented:
@kevingriffin could you post a gist of the contents of setEnvironment.fish? And maybe the untranslated bash equivalent?

kevingriffin commented:
Sure. I'll note that `env` reports the path as being set correctly, which is very strange. Can you build a system against this and see if you see the same thing?

fish: https://gist.github.com/kevingriffin/ec289a13ed6176f96e4b9f9e1d250637
bash: https://gist.github.com/kevingriffin/49116566a3c91bc99a8e01d27b2a9676

kevingriffin commented:
@bouk I've opened https://github.com/bouk/babelfish/issues/9

bouk commented:
@kevingriffin I've released version 1.0.1, please give that a shot. Thanks for taking a close look 👍

kevingriffin commented:
That was quick! It works in my testing. I'll package 1.0.1 and submit a PR for it.

kevingriffin commented:
I rebased against master to build against babelfish 1.0.1.

cole-h reviewed (approved):
Diff LGTM, and functionality seems to work fine. Thanks for your collaboration on this, @kevingriffin and @bouk!

Unless anybody finds anything wrong, I plan on merging this tomorrow morning (probably around 10 AM Pacific).

bouk reviewed (approved):
Thanks for the work on this @kevingriffin 👍🏻

Let's see how it works for people, and maybe it could be the default further down the line.

cole-h commented:
About half an hour early, but :rocket:

kevingriffin commented:
@cole-h Thank you! Excited to have this land. And thanks @bouk for the great software that made it possible.
